server/src/internalClusterTest/java/org/elasticsearch/search/functionscore/QueryRescorerIT.java

```diff
@@ -979,9 +1001,173 @@ public void testRescoreAfterCollapseRandom() throws Exception {
         });
     }
 
+    public void testRescoreWithTimeout() throws Exception {
+        // no dummy docs since merges can change scores while we run queries.
+        int numDocs = indexRandomNumbers("whitespace", -1, false);
+
+        String intToEnglish = English.intToEnglish(between(0, numDocs - 1));
+        String query = intToEnglish.split(" ")[0];
+        assertResponse(
+            prepareSearch().setSearchType(SearchType.QUERY_THEN_FETCH)
+                .setQuery(QueryBuilders.matchQuery("field1", query).operator(Operator.OR))
+                .setSize(10)
+                .addRescorer(new QueryRescorerBuilder(new TimedQueryBuilder(500)).windowSize(100))
+                .setTimeout(TimeValue.timeValueMillis(10)),
```

Comment:
Using a timeout here can be flaky. Could we use CountDownLatch and task cancellation instead? Also, we can define ScoreFunctionSpec in `SearchPlugin#getScoreFunctions()` to simulate the plugin in question.

Reply:
I think that we'd prob want both types of tests;
  1) a real timeout that verifies the timeout value is propagated correctly, and 
  2) cancellation of a request that has no explicit timeout (or a large timeout) ?

Reply:
@dnhatn I don't know how to do a countdown latch as this test could serialize. I can add an entirely new suite of tests if we want, but I thought it prudent to get a real integration test in to verify search responses are what we expect (this shows some bugs in my original approach).